src/core/library.c

```diff
@@ -30,41 +30,115 @@ QuicLibraryEvaluateSendRetryState(
+            CxPlatDispatchLockInitialize(&MsQuicLib.DatapathLock);
+            CxPlatDispatchLockInitialize(&MsQuicLib.StatelessRetryKeysLock);
+            CxPlatListInitializeHead(&MsQuicLib.Registrations);
+            CxPlatListInitializeHead(&MsQuicLib.Bindings);
+            QuicTraceRundownCallback = QuicTraceRundown;
+
+            //
+            // Try and transition to the loaded state.
+            //
+            while (NextState != (uint32_t)InterlockedCompareExchange((long*)&MsQuicLib.LoadState, (long)(NextState | LOADED_BIT), (long)NextState)) {
+                YieldProcessor();
+                NextState = (uint32_t)InterlockedOr((long*)&MsQuicLib.LoadState, 0); // No-op atomic read
+            }
+        }
+
+    } else {
+        //
+        // Busy wait until we aren't in a loading state anymore.
+        //
+        while ((NextState & LOADED_BIT) == 0) {
+            YieldProcessor();
+            NextState = (uint32_t)InterlockedOr((long*)&MsQuicLib.LoadState, 0); // No-op atomic read
+        }
+    }
 }
 
 //
-// Uninitializes global variables.
+// Uninitializes global variables if necessary.
 //
 _IRQL_requires_max_(PASSIVE_LEVEL)
 void
 MsQuicLibraryUnload(
     void
     )
 {
-    CXPLAT_FRE_ASSERT(MsQuicLib.Loaded);
-    QUIC_LIB_VERIFY(MsQuicLib.RefCount == 0);
-    QUIC_LIB_VERIFY(!MsQuicLib.InUse);
-    MsQuicLib.Loaded = FALSE;
-    CxPlatDispatchLockUninitialize(&MsQuicLib.StatelessRetryKeysLock);
-    CxPlatDispatchLockUninitialize(&MsQuicLib.DatapathLock);
-    CxPlatLockUninitialize(&MsQuicLib.Lock);
+    uint32_t NextState = (uint32_t)InterlockedDecrement((long*)&MsQuicLib.LoadState);
+    uint32_t PriorCount = (NextState & REFCOUNT_MASK) + 1;
+
+    if (PriorCount == 1) {
+        //
+        // We are responsible for unloading the library IFF another thread
+        // doesn't open the library in the meantime.
+        //
+        // Attempt to transition to the unloading state. If we fail, that
+        // means another thread has already incremented the ref count before
+        // we arrived at this instruction. Thus, we don't bother unloading
+        // the library.
+        //
+        if ((uint32_t)InterlockedCompareExchange((long*)&MsQuicLib.LoadState, UNLOADING_BIT, (long)NextState) == NextState) {
+            //
+            // We've succeeded. Unload the library.
+            //
+            QUIC_LIB_VERIFY(MsQuicLib.RefCount == 0);
+            QUIC_LIB_VERIFY(!MsQuicLib.InUse);
+            CxPlatDispatchLockUninitialize(&MsQuicLib.StatelessRetryKeysLock);
+            CxPlatDispatchLockUninitialize(&MsQuicLib.DatapathLock);
+            CxPlatLockUninitialize(&MsQuicLib.Lock);
+            CxPlatSystemUnload();
+
+            //
+            // Allow possibly spinning thread that incremented ref count in
+            // the meantime to proceed.
+            //
+            InterlockedAnd((long*)&MsQuicLib.LoadState, (long)~UNLOADING_BIT);
```

Comment:
Linux build doesn't like this:
```
/home/vsts/work/1/s/src/core/library.c:139:57: error: either cast from 'unsigned int' to 'long' is ineffective, or there is loss of precision before the conversion [bugprone-misplaced-widening-cast,-warnings-as-errors]
            InterlockedAnd((long*)&MsQuicLib.LoadState, (long)~UNLOADING_BIT)
```
Ideas?

Reply:
Hmm, that seems like a false positive. My guess is its detecting the result would be different if you casted before the not. I'd be curious if you hardcoded 0x7FFFFFFF here if it would work. As I'm not fully sure if its the left or the right side thats complaining.